multi/imxrt-multi/uart.c

```diff
@@ -86,52 +97,66 @@ static const int uartPos[] = { UART1_POS, UART2_POS, UART3_POS, UART4_POS, UART5
 enum { veridr = 0, paramr, globalr, pincfgr, baudr, statr, ctrlr, datar, matchr, modirr, fifor, waterr };
 
 
-static int uart_handleIntr(unsigned int n, void *arg)
+static inline int uart_getRXcount(uart_t *uart)
 {
-	uint32_t flags;
-	uart_t *uart = (uart_t *)arg;
+	return (*(uart->base + waterr) >> 24) & 0xff;
+}
 
-	*(uart->base + ctrlr) &= ~((1 << 27) | (1 << 26) | (1 << 25) | (1 << 23) | (1 << 21));
 
-	/* Error flags: parity, framing, noise, overrun */
-	flags = *(uart->base + statr) & (0xf << 16);
+static inline int uart_getTXcount(uart_t *uart)
+{
+	return (*(uart->base + waterr) >> 8) & 0xff;
+}
 
-	/* RX overrun: invalidate fifo */
-	if (flags & (1 << 19))
-		*(uart->base + fifor) |= 1 << 14;
 
-	*(uart->base + statr) |= flags;
+static int uart_handleIntr(unsigned int n, void *arg)
+{
+	uart_t *uart = (uart_t *)arg;
+	uint32_t status = *(uart->base + statr);
 
-	return 1;
-}
+	/* Disable interrupts, enabled in uart_intrThread */
+	*(uart->base + ctrlr) &= ~((1 << 27) | (1 << 26) | (1 << 25) | (1 << 23) | (1 << 21));
 
+	/* check for RX overrun */
+	if ((status & (1uL << 19u)) != 0u) {
+		/* invalidate fifo */
+		*(uart->base + fifor) |= 1uL << 14u;
+		uart->stat.hw_overrunCntr++;
+	}
 
-static inline int uart_getRXcount(uart_t *uart)
-{
-	return (*(uart->base + waterr) >> 24) & 0xff;
-}
+	/* Process received data */
+	while (uart_getRXcount(uart) != 0) {
+		uint8_t data = *(uart->base + datar);
+		if (lf_fifo_push(&uart->rxFifoCtx, data) == 0) {
+			uart->stat.sw_overrunCntr++;
+		}
+	}
 
+	/* Clear errors: parity, framing, noise, overrun and idle flag */
+	*(uart->base + statr) = (status & (0x1fuL << 16));
 
-static inline int uart_getTXcount(uart_t *uart)
-{
-	return (*(uart->base + waterr) >> 8) & 0xff;
+	/* reschedule */
```

Comment:
Nitpick - not exactly, wakeup registered cond (uart->cond)

Reply:
Right, thanks.